test/acceptance/tests/hcp/hcp_test.go

```diff
@@ -155,6 +150,402 @@ func TestHCP(t *testing.T) {
+	retry.RunWith(timeouts["mesh-task"], t, func(r *retry.R) {
+		curlOut, err := task.ExecuteCommand("basic", `/bin/sh -c "curl localhost:1234"`)
+		r.Check(err)
+		if !strings.Contains(curlOut, expected) {
+			r.Errorf("response was unexpected: %q", curlOut)
+		}
+	})
+}
+
+func upsertIntention(t *testing.T, consulClient *api.Client, action api.IntentionAction, src, dst *helpers.MeshTask) {
+	retry.RunWith(timeouts["consul"], t, func(r *retry.R) {
+		_, _, err := consulClient.ConfigEntries().Set(&api.ServiceIntentionsConfigEntry{
+			Kind:      api.ServiceIntentions,
+			Name:      dst.Name,
+			Partition: dst.Partition,
+			Namespace: dst.Namespace,
+			Sources: []*api.SourceIntention{
+				&api.SourceIntention{
+					Name:      src.Name,
+					Partition: src.Partition,
+					Namespace: src.Namespace,
+					Action:    action,
+				},
+			},
+		}, dst.WriteOpts())
+		r.Check(err)
+	})
+}
+
+func deleteIntention(t *testing.T, consulClient *api.Client, dst *helpers.MeshTask) {
+	retry.RunWith(timeouts["consul"], t, func(r *retry.R) {
+		_, err := consulClient.ConfigEntries().Delete(api.ServiceIntentions, dst.Name, dst.WriteOpts())
+		r.Check(err)
+	})
+}
+
+func upsertExportedServices(t *testing.T, consulClient *api.Client, src, dst *helpers.MeshTask) {
+	retry.RunWith(timeouts["consul"], t, func(r *retry.R) {
+		_, _, err := consulClient.ConfigEntries().Set(&api.ExportedServicesConfigEntry{
+			Name:      dst.Partition,
+			Partition: dst.Partition,
+			Services: []api.ExportedService{{
+				Name:      dst.Name,
+				Namespace: dst.Namespace,
+				Consumers: []api.ServiceConsumer{{Partition: src.Partition}},
+			}},
+		}, dst.WriteOpts())
+		r.Check(err)
+	})
+}
+
+func deleteExportedServices(t *testing.T, consulClient *api.Client, dst *helpers.MeshTask) {
+	retry.RunWith(timeouts["consul"], t, func(r *retry.R) {
+		_, err := consulClient.ConfigEntries().Delete(api.ExportedServices, dst.Partition, dst.WriteOpts())
+		r.Check(err)
+	})
+}
+
+func consulClient(t *testing.T, addr, token string) (*api.Client, ConsulState, error) {
+	cfg := api.DefaultConfig()
+	cfg.Address = addr
+	cfg.Token = token
+	client, err := api.NewClient(cfg)
+	if err != nil {
+		return nil, ConsulState{}, fmt.Errorf("failed to create Consul client: %w", err)
+	}
+	state, err := recordConsulState(t, client)
+	if err != nil {
+		return nil, ConsulState{}, fmt.Errorf("failed to establish Consul state: %w", err)
+	}
+	return client, state, err
+}
+
+type ConsulState struct {
+	Partitions map[string]PartitionState
+}
+
+type PartitionState struct {
+	Name       string
+	Namespaces map[string]NamespaceState
+}
+
+type NamespaceState struct {
+	Name        string
+	Tokens      map[string]struct{}
+	Policies    map[string]struct{}
+	Roles       map[string]struct{}
+	AuthMethods map[string]struct{}
+}
+
+func recordConsulState(t *testing.T, consul *api.Client) (ConsulState, error) {
+	// TODO: not sure how to handle OSS vs Enterprise cases.. these tests currently use Enterprise.
```

Comment:
This is much better than before. So, thanks a lot for making the enterprise cases significantly better 😄

Reply:
My pleasure :)